File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter20/intermediate/IntermediateEcosystem.kt
package com.devlokos.masteringkotlin.chapters.chapter20.intermediate

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * Capítulo 20 — Ecosistema y Bibliotecas.
 * Nivel INTERMEDIO: DateUtils — ayudantes de fecha estilo biblioteca.
 */
fun main() {
    runIntermediateEcosystemExamples()
}

object DateUtils {
    private val formatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    fun formatDate(timestamp: Long): String = formatter.format(Date(timestamp))

    fun daysBetween(fromMs: Long, toMs: Long): Long =
        TimeUnit.MILLISECONDS.toDays(toMs - fromMs)
}

fun runIntermediateEcosystemExamples() {
    println("DateUtils.formatDate(0): ${DateUtils.formatDate(0)}")
    val day = 86_400_000L  // 1 day in ms
    println("DateUtils.daysBetween(0, 2*day): ${DateUtils.daysBetween(0, 2 * day)}")
}

File: MasteringSwift/MasteringSwift/chapters/chapter20/intermediate/IntermediateEcosystem.swift
//
//  IntermediateEcosystem.swift
//  MasteringSwift
//
//  Capítulo 20 — Ecosistema y bibliotecas.
//  Nivel INTERMEDIO: DateUtils — ayudantes de fecha estilo biblioteca.
//

import Foundation

enum DateUtils {
    private static let formatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    static func formatDate(timestamp: TimeInterval) -> String {
        formatter.string(from: Date(timeIntervalSince1970: timestamp))
    }

    static func daysBetween(from: Date, to: Date) -> Int {
        Calendar.current.dateComponents([.day], from: from, to: to).day ?? 0
    }
}

func runIntermediateEcosystemExamples() {
    print("DateUtils.formatDate(0): \(DateUtils.formatDate(timestamp: 0))")
    let from = Date(timeIntervalSince1970: 0)
    let to = Date(timeIntervalSince1970: 2 * 86400)
    print("DateUtils.daysBetween(0, 2 days): \(DateUtils.daysBetween(from: from, to: to))")
}
